### PR TITLE
Use react-redux instead of custom Provider

### PR DESCRIPTION
**Pull request by jseppi**
Here's a first round of refactoring our React app to use `react-redux` (ref #1000) in place of the custom `Provider` component we had.

Summary of changes:
- Remove our `CustomProvider` and replace it with `react-redux`'s `Provider`
- Use `react-redux`'s `connect` method to map our store's `state` to `props` of components.
- Add `redux-devtools-extension` so we can use the [Redux Devtools browser extension](http://extension.remotedev.io/)
- Adjust tests to operate on component `props` instead of store `state`
- Add additional tests for `App` component

### CODE REVIEW

**Comment by jseppi** [comment]
@jmhooper: I think this is ready for review. Pending successful review, I'll `rebase` before a merge because this has some particularly ugly commit history as I was figuring things out.

**Comment by jseppi** [comment]
Please put it through the ringer though since this is a somewhat fundamental change :)

File: frontend/components/app.js [hunk 1]
```diff
@@ -1,35 +1,16 @@
...
-
-class App extends React.Component {
-  constructor(props, context) {
-    super(props, context);
-    this.state = this.getStateFromStore();
-  }
-
+export class App extends React.Component {
   componentWillReceiveProps(nextProps) {
-    const state = this.getStateFromStore();
```
Comment (jmhooper) [inline frontend/components/app.js]: I'll be happy when we can move everything over to our notification lib and all of this code can go away.

File: frontend/components/app.js [hunk 2]
```diff
@@ -46,48 +27,75 @@ class App extends React.Component {
   }
 
   render() {
-    const { children } = this.props;
-    const storeState = this.state;
+    const { user, children, notifications } = this.props;
 
-    if (storeState.user.isLoading) {
+    if (!user || user.isLoading) {
```
Comment (jmhooper) [inline frontend/components/app.js]: The [user reducer sets the state to `false` if the user fetch fails](https://github.com/18F/federalist/blob/staging/frontend/reducers/user.js#L17-L19). This means that for an unauthenticated user, `this.props.user` will be false.

When there is an authed user, the `App` component renders the `Home` component at its index. The `Home` component [redirects to the site list when it finds a logged in user](https://github.com/18F/federalist/blob/staging/frontend/components/home.js#L18). 

But, without an authed user, the `App` component expects to render the `Home` component, and for the `Home` component to display our fancy marketing content. It looks like this change breaks that by having the App component perpetually render the loading indicator if the user is not logged in.

Comment (jseppi) [inline frontend/components/app.js]: 👀

Comment (jseppi) [inline frontend/components/app.js]: oof, good catch. I think this is another good case for https://github.com/18F/federalist/issues/958

Comment (jseppi) [inline frontend/components/app.js]: I'll fix it 'as-is' here though, to keep this PR from ballooning

**Comment by jseppi** [comment]
@jmhooper - I fixed the bug you found and added tests to cover those conditions

File: frontend/components/app.js [hunk 3]
```diff
@@ -46,48 +27,75 @@ class App extends React.Component {
   }
 
   render() {
-    const { children } = this.props;
-    const storeState = this.state;
+    const { user, children, notifications } = this.props;
 
-    if (storeState.user.isLoading) {
+    if (user.isLoading) {
```
Comment (jmhooper) [inline frontend/components/app.js]: Safe to assume that `user` is never undefined? I can't seem to make it break so I think so.

Comment (jseppi) [inline frontend/components/app.js]: the user reducer sets its initial state to so it will never be undefined

**Comment by jseppi** [comment]
I felt the same thing about the apparent quickness with this change!

I'll rebase this PRs commit history to be cleaner, then :shipit:

**Comment by wslack** [comment]
it does feel snappier.

**Pull request merged**